flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepoint.java

```diff
@@ -0,0 +1,317 @@
+import org.apache.flink.runtime.scheduler.stopwithsavepoint.StopWithSavepointOperationHandlerImpl;
+import org.apache.flink.runtime.scheduler.stopwithsavepoint.StopWithSavepointOperationManager;
+import org.apache.flink.util.Preconditions;
+
+import org.slf4j.Logger;
+
+import javax.annotation.Nullable;
+
+import java.time.Duration;
+import java.util.Collection;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * When a "stop with savepoint" operation (wait until savepoint has been created, then cancel job)
+ * is triggered on the {@link Executing} state, we transition into this state. This state is
+ * delegating the tracking of the stop with savepoint operation to the {@link
+ * StopWithSavepointOperationManager}, which is shared with {@link SchedulerBase}.
+ */
+class StopWithSavepoint extends StateWithExecutionGraph implements StopWithSavepointOperations {
+
+    private final CompletableFuture<String> operationCompletionFuture;
+    private final Context context;
+    private final ClassLoader userCodeClassLoader;
+    private final StopWithSavepointOperationManager stopWithSavepointOperationManager;
+
+    StopWithSavepoint(
+            Context context,
+            ExecutionGraph executionGraph,
+            ExecutionGraphHandler executionGraphHandler,
+            OperatorCoordinatorHandler operatorCoordinatorHandler,
+            Logger logger,
+            ClassLoader userCodeClassLoader,
+            @Nullable String targetDirectory,
+            boolean terminate) {
+        super(context, executionGraph, executionGraphHandler, operatorCoordinatorHandler, logger);
+        this.context = context;
+        this.userCodeClassLoader = userCodeClassLoader;
+
+        // to ensure that all disjoint subgraphs of a job finish successfully on savepoint creation,
+        // we track the job termination via all execution termination futures (FLINK-21030).
+        final CompletableFuture<Collection<ExecutionState>> executionTerminationsFuture =
+                SchedulerUtils.getCombinedExecutionTerminationFuture(executionGraph);
+
+        stopWithSavepointOperationManager =
+                new StopWithSavepointOperationManager(
+                        this,
+                        new StopWithSavepointOperationHandlerImpl(
+                                executionGraph.getJobID(), context, this, logger));
+
+        this.operationCompletionFuture =
+                stopWithSavepointOperationManager.trackStopWithSavepoint(
+                        terminate,
+                        targetDirectory,
+                        executionTerminationsFuture,
+                        context.getMainThreadExecutor());
+    }
+
+    @Override
+    public void onLeave(Class<? extends State> newState) {
+        // we need to delay the execution because this may trigger a state transition of the
+        // AdaptiveScheduler.
+        context.getMainThreadExecutor()
+                .execute(
```

Comment:
I think the problem you are trying to work around with this construct is that the `stopWithSavepointOperationManager.abortOperation` can trigger a `handleGlobalFailure` if it happens after the savepoint has been created. With this change we will trigger a global failure on the next state. I think this is not correct as we should only leave the `StopWithSavepoint` state if the stop with savepoint operation has been completed (one way or the other).